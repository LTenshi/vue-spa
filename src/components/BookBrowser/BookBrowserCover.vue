<template>
  <tr>
    <th class="imageHeader">
        <img :src="bookCover" @click="openDetails" class="imageContainer">
    </th>
    <th class="bookName">
        <div>{{bookName}}</div>
        <div class="bookAuthor">{{bookAuthor}}</div>
    </th>
  </tr>
</template>

<script>
export default {
    props: {
        bookName: {
            type: String,
            required: true,
        },
        bookAuthor: {
            type: String,
            required: true,
        },
        bookCover: {
            type: String,
            required: true,
        },
    },
    methods: {
        openDetails() {
            this.$emit("openDetails");
        }
    }
}
</script>

<style>
.bookName, .bookAuthor {
    font-size: 13pt;
}

.bookName {
    vertical-align: top;
    padding: 20px;
    text-align: left;
}

.bookAuthor {
    font-weight: 400;
    color: rgb(175, 175, 175);
}

.imageContainer {
    max-width: 140px;
}
</style>