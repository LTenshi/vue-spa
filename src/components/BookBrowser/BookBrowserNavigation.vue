<template>
    <th class="container" colspan="4">
        <button @click="moveBack" :style="isBackPossible" :disabled="isBackPossible">&#10094;</button>
        Page {{currPage}} of {{maxPages}} 
        <button @click="moveForward" :style="isForwardPossible" :disabled="isForwardPossible">&#10095;</button>
    </th>
</template>

<script>
export default {
 props: {
    maxPages: {
      type: Number,
      required: true
    },
    currPage: {
        type: Number,
        required: true,
    }
  },
  methods: {
    moveBack() {
        this.$emit("navigationEvent", this.currPage - 1);
    },
    moveForward() {
        this.$emit("navigationEvent", this.currPage + 1);
    }
  },
  computed: {
    isBackPossible() {
        return this.currPage === 1 ? { color: "lightgrey", cursor: "unset" } : false
    },
    isForwardPossible() {
        return this.currPage === this.maxPages ? { color: "lightgrey", cursor: "unset" } : false
    }
  }
}
</script>

<style>
.container {
    width: 100%;
}

button {
	background: none;
    color: grey;
	border: none;
	font: inherit;
	cursor: pointer;
}
</style>