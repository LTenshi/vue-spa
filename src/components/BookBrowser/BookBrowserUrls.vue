<template>
  <th>
    <a v-for="link in bookLinks" :key="link.name" :href="link.url" class="bookLink">{{link.name}}<br></a>
  </th>
</template>

<script>
export default {
  props: {
    bookLinks: {
      type: Array,
      default: () => []
    }
  }
}
</script>

<style>
.bookLink {
  color: rgb(136, 221, 0);
  text-decoration: none;
}

.bookLink > a:link {
  text-decoration: none;
}

.bookLink > a:visited {
  text-decoration: none;
}

.bookLink > a:hover {
  text-decoration: none;
}

.bookLink > a:active {
  text-decoration: none;
}
</style>