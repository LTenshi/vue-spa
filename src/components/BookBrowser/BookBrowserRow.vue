<template>
  <tr>
    <book-browser-cover 
        :bookCover="bookCover" 
        :bookName="bookName" 
        :bookAuthor="bookAuthor"
        @openDetails="openDetails"/>
    <th class="fontStyle">
        {{this.bookYear}}
    </th>
    <th class="fontStyle">
        {{this.bookRating}}/10
    </th>
    <book-browser-urls :bookLinks="bookLinks"/>
  </tr>
  <th v-if="detailsExpanded" colspan="4" class="detailsPanel" v-html="this.bookDetails"></th>
</template>

<script>
import BookBrowserCover from './BookBrowserCover.vue';
import BookBrowserUrls from './BookBrowserUrls.vue';
import showdown from "showdown";
export default {
    components: { BookBrowserCover, BookBrowserUrls },
    data() {
        return {
            detailsExpanded: false
        }
    },
    props: {
        book: {
            type: Object,
            required: true,
        },
    },
    computed: {
        bookCover() {
            return this.book.coverUrl;
        },
        bookName() {
            return this.book.name;
        },
        bookId() {
            return this.book.id;
        },
        bookAuthor() {
            return this.book.author;
        },
        bookYear() {
            return this.book.year;
        },
        bookRating() {
            return this.book.rating;
        },
        bookLinks() {
            return this.book.links;
        },
        bookDetails() {
            var converter = new showdown.Converter();
            return converter.makeHtml(this.book.details);
        }
    },
    methods: {
        openDetails() {
            this.detailsExpanded = !this.detailsExpanded;
        }
    }
}
</script>

<style>
.detailsPanel {
    font-weight: 450;
    text-align: initial;
}

.fontStyle {
    font-weight: 450;
}
</style>