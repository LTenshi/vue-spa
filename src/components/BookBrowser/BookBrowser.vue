<template>
  <table id="bookbrowser">
    <thead>
      <h3>Most popular Books of All time</h3>
    </thead>
    <tbody class="rowContainer">
      <tr>
        <th v-for="value in columns" :key="value.title" class="tHeader" :style="headerStyle(value.style)">
          {{capitalise(value.title)}}
        </th>
      </tr>
      <book-browser-row 
        v-for="book in books" 
        :key="book.id"
        :book="book">
      </book-browser-row>
    </tbody>
    <tfoot>
      <book-browser-navigation :maxPages="maxPages" :currPage="currPage" @navigationEvent="navigationEventHandler"/>
    </tfoot>
  </table>
</template>

<script>
import BookBrowserRow from './BookBrowserRow.vue'
import BookBrowserNavigation from "./BookBrowserNavigation.vue"
export default {
  components: { BookBrowserRow, BookBrowserNavigation },
  data: function() {
    return {
      columns: [
        { title: "Title", style: { width: "60", textAlign: "initial" } },
        { title: "Published", style: { width: "10" } },
        { title: "Rating", style:{ width: "10" } },
        { title: "Buy On", style: { width: "10" } },
      ]
    }
  },
  props: {
    books: {
      type: Array,
      required: true,
      default: () => []
    },
    maxPages: {
      type: Number,
      required: true
    },
    currPage: {
      type: Number,
      required: true
    }
  },
  methods: {
    capitalise(value) {
      return value.toUpperCase();
    },
    headerStyle(style) {
      let styleReturn = {};
      styleReturn.width = style.width + "%";
      if (style.textAlign) {
        styleReturn.textAlign = style.textAlign;
      }
      return styleReturn
    },
    navigationEventHandler(navEvent) {
      this.$emit("navigationEvent", navEvent);
    }
  },
}
</script>

<style>

#bookbrowser {
  width: 100%;
  background-color: white;
  border-radius: 5px;
  padding: 10px;
}

.rowContainer > tr:nth-child(even) {
  background-color: rgb(245, 245, 245);
}

.tHeader {
  color: rgb(175, 175, 175);
}

</style>