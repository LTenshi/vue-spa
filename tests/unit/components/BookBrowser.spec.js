import BookBrowser from "../../../src/components/BookBrowser/BookBrowser";
import BookBrowserRow from "../../../src/components/BookBrowser/BookBrowserRow";
import BookBrowserNavigation from "../../../src/components/BookBrowser/BookBrowserNavigation";
import { mount } from "@vue/test-utils";

describe("Book Browser", () => {
    let wrapper;
    let BookBrowserRowComponent;
    let BookBrowserNavigationComponent;

    beforeEach(() => {
        wrapper = mount(BookBrowser, {
            props: {
                books: [
                    {
                        "id": "0",
                        "coverUrl": "https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fcdn2.penguin.com.au%2Fcovers%2Foriginal%2F9780141956794.jpg&f=1&nofb=1",
                        "name": "In Search of Lost Time",
                        "author": "Marcel Proust",
                        "year": "1913",
                        "rating": "9.8",
                        "links": [
                        { "name": "Amazon", "url": "https://www.amazon.co.uk" },
                        { "name": "iBooks", "url": "https://www.apple.com/uk/apple-books/" },
                        { "name": "Play Store", "url": "https://play.google.com/store/books" }
                        ],
                        "details": "**In Search of Lost Time** (French: Ã€ la recherche du temps perdu), first translated into English as Remembrance of Things Past, and sometimes referred to in French as La Recherche (The Search), is a novel in seven volumes by French author Marcel Proust.This early 20th-century work is his most prominent, known both for its length and its theme of involuntary memory."
                    },
                    {
                        "id": "1",
                        "coverUrl": "https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fuploads7.wikiart.org%2Fimages%2Fhonore-daumier%2Fdon-quixote-and-sancho-panza-going-to-the-wedding-gamaches.jpg&f=1&nofb=1",
                        "name": "Don Quixote",
                        "author": "Miguel de Cervantes",
                        "year": "1605",
                        "rating": "9.7",
                        "links": [
                        { "name": "Amazon", "url": "https://www.amazon.co.uk" }
                        ],
                        "details": "**Don Quixote** is a Spanish epic novel by Miguel de Cervantes. Originally published in two parts, in 1605 and 1615, its full title is The Ingenious Gentleman Don Quixote of La Mancha or, in Spanish, El ingenioso hidalgo don Quijote de la Mancha."
                    }
                ],
                maxPages: 1,
                currPage: 1
            }
        });
        BookBrowserRowComponent = wrapper.findAllComponents(BookBrowserRow);
        BookBrowserNavigationComponent = wrapper.findComponent(BookBrowserNavigation);
    })

    it("Should match the snapshot", () => {
        expect(wrapper.element).toMatchSnapshot();
    })

    it('Should render the book browser rows', () => {
        expect(BookBrowserRowComponent.length).toBe(2);
    });

    it('Should render the book browser rows', () => {
        expect(BookBrowserNavigationComponent.exists()).toBe(true);
    });

    it('Should have no ability to navigate', () => {
        expect(BookBrowserNavigationComponent.vm.isBackPossible).toStrictEqual({"color": "lightgrey", "cursor": "unset"});
        expect(BookBrowserNavigationComponent.vm.isForwardPossible).toStrictEqual({"color": "lightgrey", "cursor": "unset"});
    });

    it('Should contain capitalised headers', () => {
        expect(wrapper.html()).toContain("TITLE");
        expect(wrapper.html()).toContain("PUBLISHED");
        expect(wrapper.html()).toContain("RATING");
        expect(wrapper.html()).toContain("BUY ON");
    });

});