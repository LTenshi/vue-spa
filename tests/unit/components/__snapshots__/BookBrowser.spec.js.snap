// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Browser Should match the snapshot 1`] = `
<table
  id="bookbrowser"
>
  <thead>
    <h3>
      Most popular Books of All time
    </h3>
  </thead>
  <tbody
    class="rowContainer"
  >
    <tr>
      
      <th
        class="tHeader"
        style="width: 60%; text-align: initial;"
      >
        TITLE
      </th>
      <th
        class="tHeader"
        style="width: 10%;"
      >
        PUBLISHED
      </th>
      <th
        class="tHeader"
        style="width: 10%;"
      >
        RATING
      </th>
      <th
        class="tHeader"
        style="width: 10%;"
      >
        BUY ON
      </th>
      
    </tr>
    
    
    <tr>
      <tr>
        <th
          class="imageHeader"
        >
          <img
            class="imageContainer"
            src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fcdn2.penguin.com.au%2Fcovers%2Foriginal%2F9780141956794.jpg&f=1&nofb=1"
          />
        </th>
        <th
          class="bookName"
        >
          <div>
            In Search of Lost Time
          </div>
          <div
            class="bookAuthor"
          >
            Marcel Proust
          </div>
        </th>
      </tr>
      <th
        class="fontStyle"
      >
        1913
      </th>
      <th
        class="fontStyle"
      >
        9.8/10 
      </th>
      <th>
        
        <a
          class="bookLink"
          href="https://www.amazon.co.uk"
        >
          Amazon
          <br />
        </a>
        <a
          class="bookLink"
          href="https://www.apple.com/uk/apple-books/"
        >
          iBooks
          <br />
        </a>
        <a
          class="bookLink"
          href="https://play.google.com/store/books"
        >
          Play Store
          <br />
        </a>
        
      </th>
    </tr>
    <!--v-if-->
    
    
    <tr>
      <tr>
        <th
          class="imageHeader"
        >
          <img
            class="imageContainer"
            src="https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fuploads7.wikiart.org%2Fimages%2Fhonore-daumier%2Fdon-quixote-and-sancho-panza-going-to-the-wedding-gamaches.jpg&f=1&nofb=1"
          />
        </th>
        <th
          class="bookName"
        >
          <div>
            Don Quixote
          </div>
          <div
            class="bookAuthor"
          >
            Miguel de Cervantes
          </div>
        </th>
      </tr>
      <th
        class="fontStyle"
      >
        1605
      </th>
      <th
        class="fontStyle"
      >
        9.7/10 
      </th>
      <th>
        
        <a
          class="bookLink"
          href="https://www.amazon.co.uk"
        >
          Amazon
          <br />
        </a>
        
      </th>
    </tr>
    <!--v-if-->
    
    
  </tbody>
  <tfoot>
    <th
      class="container"
      colspan="4"
    >
      <button
        disabled=""
        style="color: lightgrey; cursor: unset;"
      >
        ❮
      </button>
       Page 1 of 1 
      <button
        disabled=""
        style="color: lightgrey; cursor: unset;"
      >
        ❯
      </button>
    </th>
  </tfoot>
</table>
`;
